lib/Doctrine/DBAL/Query/QueryBuilder.php

```diff
@@ -1012,59 +1029,107 @@ public function orderBy(string $sort, ?string $order = null)
      */
     public function addOrderBy(string $sort, ?string $order = null)
     {
-        return $this->add('orderBy', $sort . ' ' . (! $order ? 'ASC' : $order), true);
+        $this->queryParts->orderBy[] = $sort . ' ' . (! $order ? 'ASC' : $order);
+
+        $this->state = self::STATE_DIRTY;
+
+        return $this;
     }
 
     /**
-     * Gets a query part by its name.
+     * Gets all query parts.
      *
-     * @return mixed
+     * @internal For tests only.
      */
-    public function getQueryPart(string $queryPartName)
+    public function getQueryParts() : QueryParts
     {
-        return $this->sqlParts[$queryPartName];
+        return $this->queryParts;
     }
 
-    /**
-     * Gets all query parts.
-     *
-     * @return array<string, mixed>
-     */
-    public function getQueryParts() : array
+    public function resetSelect() : QueryBuilder
     {
-        return $this->sqlParts;
+        $this->queryParts->select = [];
+        $this->state              = self::STATE_DIRTY;
+
+        return $this;
     }
 
-    /**
-     * Resets SQL parts.
-     *
-     * @param array<int, string>|null $queryPartNames
-     *
-     * @return $this This QueryBuilder instance.
-     */
-    public function resetQueryParts(?array $queryPartNames = null) : self
+    public function resetDistinct() : QueryBuilder
     {
-        if ($queryPartNames === null) {
-            $queryPartNames = array_keys($this->sqlParts);
-        }
+        $this->queryParts->distinct = false;
+        $this->state                = self::STATE_DIRTY;
 
-        foreach ($queryPartNames as $queryPartName) {
-            $this->resetQueryPart($queryPartName);
-        }
+        return $this;
+    }
+
+    public function resetFrom() : QueryBuilder
+    {
+        $this->queryParts->from = [];
+        $this->state            = self::STATE_DIRTY;
 
         return $this;
     }
 
-    /**
-     * Resets a single SQL part.
-     *
-     * @return $this This QueryBuilder instance.
-     */
-    public function resetQueryPart(string $queryPartName) : self
+    public function resetTable() : QueryBuilder
     {
-        $this->sqlParts[$queryPartName] = self::SQL_PARTS_DEFAULTS[$queryPartName];
+        $this->queryParts->table = null;
+        $this->state             = self::STATE_DIRTY;
 
-        $this->state = self::STATE_DIRTY;
+        return $this;
+    }
+
+    public function resetJoin() : QueryBuilder
```

Comment:
**Should we introduce all these `resetXXX()` methods?** I've added them to replace `resetQueryPart[s]()`; only [2 of them](https://github.com/doctrine/dbal/blob/dbd94c9d591e3d6f980f3a077cfa5b64c600b6d7/tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php#L564-L565) are used in tests at the moment.

Reply:
No. In the cases I'm aware of, the `reset*()` methods indicate poor design. E.g. one method builds a query with `JOIN`s and `ORDER BY`, then the other one resets the `SELECT` and `ORDER BY` parts and adds `COUNT(*)`.

Instead, only the needed parts of the query should be built for each case.

Let's keep the existing methods for now but not add any new.

Reply:
> Let's keep the existing methods for now but not add any new.

Which ones? The existing methods are generic `resetQueryPart[s]()`, which we're removing anyway. Should I remove all `resetXXX()` methods then?

Reply:
Oh, I thought we had some methods like `resetJoin()`. Let's not introduce any.

Reply:
Done in 6369a4a2cdcd3c240c4a9e09fe5fa5473ffc0ac0.